### PR TITLE
Optimize key creation routines.

### PR DESCRIPTION
**Pull request by petermattis**
Remove usage of {keys,roachpb}.MakeKey from common paths. Moved test to
testutils.MakeKey to prevent accidental reintroduction. The key creation
routines in `keys` now preallocate a buffer of the correct size.

```
name                   old time/op    new time/op    delta
Insert1_Cockroach-8       379µs ± 2%     370µs ± 0%  -2.55%   (p=0.000 n=10+9)
Insert10_Cockroach-8      775µs ± 2%     751µs ± 2%  -3.05%  (p=0.000 n=10+10)
Insert100_Cockroach-8    4.17ms ± 1%    4.07ms ± 0%  -2.22%   (p=0.000 n=10+9)
Update1_Cockroach-8       857µs ± 1%     843µs ± 1%  -1.64%  (p=0.000 n=10+10)
Update10_Cockroach-8     4.65ms ± 0%    4.59ms ± 0%  -1.31%  (p=0.000 n=10+10)
Update100_Cockroach-8    41.0ms ± 1%    40.3ms ± 1%  -1.69%  (p=0.000 n=10+10)
Delete1_Cockroach-8      1.18ms ± 0%    1.18ms ± 1%    ~      (p=0.887 n=7+10)
Delete10_Cockroach-8     1.93ms ± 1%    1.90ms ± 1%  -1.49%  (p=0.001 n=10+10)
Delete100_Cockroach-8    11.0ms ± 2%    10.7ms ± 0%  -3.21%    (p=0.000 n=8+8)
Scan1_Cockroach-8         181µs ± 1%     180µs ± 0%  -0.55%   (p=0.028 n=10+9)
Scan10_Cockroach-8        223µs ± 1%     221µs ± 1%  -0.87%  (p=0.001 n=10+10)
Scan100_Cockroach-8       614µs ± 2%     597µs ± 3%  -2.78%  (p=0.019 n=10+10)

name                   old allocs/op  new allocs/op  delta
Insert1_Cockroach-8         474 ± 0%       438 ± 0%  -7.59%   (p=0.000 n=9+10)
Insert10_Cockroach-8      1.12k ± 0%     1.08k ± 0%  -3.28%  (p=0.000 n=10+10)
Insert100_Cockroach-8     7.31k ± 0%     7.27k ± 0%  -0.52%   (p=0.000 n=9+10)
Update1_Cockroach-8       1.06k ± 0%     0.99k ± 0%  -6.50%  (p=0.000 n=10+10)
Update10_Cockroach-8      7.11k ± 0%     6.75k ± 0%  -5.15%    (p=0.000 n=8+7)
Update100_Cockroach-8     67.5k ± 0%     64.1k ± 0%  -4.95%   (p=0.000 n=10+9)
Delete1_Cockroach-8       1.01k ± 0%     0.95k ± 0%  -6.46%  (p=0.000 n=10+10)
Delete10_Cockroach-8      2.49k ± 0%     2.42k ± 0%  -2.67%   (p=0.000 n=10+9)
Delete100_Cockroach-8     16.6k ± 0%     16.5k ± 0%  -0.39%   (p=0.000 n=10+8)
Scan1_Cockroach-8           285 ± 0%       280 ± 0%  -1.86%  (p=0.000 n=10+10)
Scan10_Cockroach-8          376 ± 0%       370 ± 0%  -1.46%   (p=0.000 n=10+9)
Scan100_Cockroach-8       1.11k ± 0%     1.11k ± 0%  -0.42%   (p=0.000 n=10+8)
```

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4173)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 5 of 15 files at r1.
Review status: 5 of 15 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[keys/keys.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-K9o9ocslL8wlsChjvLC:-K9o9ocslL8wlsChjvLD:-392313706) ([raw file](https://github.com/cockroachdb/cockroach/blob/89dd4bb4202712715403ea469a430349c85bd3dd/keys/keys.go#L33)):*
Isn't this all just because `roachpb.MakeKey` is strangely inefficient?

Isn't it enough to apply this diff:
```diff
diff --git a/roachpb/data.go b/roachpb/data.go
index 723faa7..86664ab 100644
--- a/roachpb/data.go
+++ b/roachpb/data.go
@@ -112,11 +112,7 @@ type Key []byte
 // MakeKey makes a new key which is the concatenation of the
 // given inputs, in order.
 func MakeKey(keys ...[]byte) []byte {
-	byteSlices := make([][]byte, len(keys))
-	for i, k := range keys {
-		byteSlices[i] = []byte(k)
-	}
-	return bytes.Join(byteSlices, nil)
+	return bytes.Join(keys, nil)
 }
 
 // BytesNext returns the next possible byte by appending an \x00.

```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173)*

**Comment by petermattis** [comment]
Review status: 5 of 15 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[keys/keys.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-K9o9ocslL8wlsChjvLC:-K9oAIXbVddzwg8GUou-:-1075757096) ([raw file](https://github.com/cockroachdb/cockroach/blob/89dd4bb4202712715403ea469a430349c85bd3dd/keys/keys.go#L33)):*
No, that gets part of the way there, but creating a single buffer of the correct length is better when we toss in the `encoding.Encode*` routines as well.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173)*

**Comment by tamird** [comment]
Review status: 5 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[keys/keys.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-K9o9ocslL8wlsChjvLC:-K9oDeekgdYApSnkKciM:-170673961) ([raw file](https://github.com/cockroachdb/cockroach/blob/89dd4bb4202712715403ea469a430349c85bd3dd/keys/keys.go#L33)):*
it's really hard to read this diff - there are a bunch of changes lumped into one commit: unexporting makeKey, moving stuff to testutils. can you break things up?

---

*[keys/keys.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-K9oDztl0kr5UtQrz9Q7:-K9oDztl0kr5UtQrz9Q8:-234895084) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c875ca75d04c617e1c21bc77cacdeb349510a2d/keys/keys.go#L37)):*
i dont know what i'm missing here - this looks exactly the same as bytes.Join

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173)*

**Comment by petermattis** [comment]
Review status: 5 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[keys/keys.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-K9o9ocslL8wlsChjvLC:-K9oEuhnXBf6hdXY9-G5:-40219114) ([raw file](https://github.com/cockroachdb/cockroach/blob/89dd4bb4202712715403ea469a430349c85bd3dd/keys/keys.go#L33)):*
The changes are all related and, outside of this file, mostly mechanical. I'm not aware of any requirement to place potentially separable changes in separate commits. Certainly, some people do that, but that's their prerogative. 

I just took another scan through the diff and it still looks mostly mechanical to me (again, outside of  this file). Perhaps take another look on Monday with fresher eyes.

---

*[keys/keys.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-K9oDztl0kr5UtQrz9Q7:-K9oET14D2EzV8pQUnh2:1196025850) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c875ca75d04c617e1c21bc77cacdeb349510a2d/keys/keys.go#L37)):*
`bytes.Join` requires passing a `[][]byte`. Where did that come from? And this is the simplest case. See `StoreStatusKey` for an example where `encoding.EncodeUvarintAscending` is used which has to allocate if it isn't passed an appropriately sized buffer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173)*

**Comment by tamird** [comment]
LGTM

---

Reviewed 10 of 15 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[keys/keys.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-K9o9ocslL8wlsChjvLC:-K9oIn0O2Xiuwoxj2Gyr:-665118321) ([raw file](https://github.com/cockroachdb/cockroach/blob/89dd4bb4202712715403ea469a430349c85bd3dd/keys/keys.go#L33)):*
There's no requirement, it's just hard to review because I can't skip over all the mechanical changes.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4173#-:-K9oJXJ4Ryxelp_OIFla:676834233)*

File: storage/addressing.go [hunk 1]
```diff
@@ -62,6 +62,13 @@ func updateRangeAddressing(b *client.Batch, desc *roachpb.RangeDescriptor) error
 	return rangeAddressing(b, desc, putMeta)
 }
 
+func makeMetaKey(prefix roachpb.Key, key roachpb.RKey) roachpb.Key {
```
Comment (tschottdorf) [inline storage/addressing.go]: shouldn't this live in `keys`? I think it's worthwhile not to create ad-hoc keys anywhere in the codebase.

File: ts/keys.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ import (
 // 		slot := (timestamp / keyDuration) // integer division
 var (
 	// keyDataPrefix is the key prefix for time series data keys.
-	keyDataPrefix = roachpb.Key(roachpb.MakeKey(keys.SystemPrefix, roachpb.Key("tsd")))
+	keyDataPrefix = roachpb.Key(string(keys.SystemPrefix) + "tsd")
```
Comment (tschottdorf) [inline ts/keys.go]: ditto

File: storage/addressing.go (hunk 1 above)
Comment (petermattis) [inline storage/addressing.go]: Probably. I was doing the minimal change here. Let me take a look at moving it to keys.

File: ts/keys.go (hunk 2 above)
Comment (petermattis) [inline ts/keys.go]: Yeah, this definitely belongs in `keys`.

**Pull request merged**